ledger/vm/vm.go

```diff
@@ -68,6 +71,177 @@ func TransferTheta(db StateDB, sender, recipient common.Address, amount *big.Int
+		return
+	}
+	holderSig, err = crypto.SignatureFromBytes(summary[164:])
+	if err != nil {
+		return
+	}
+
+	ok = true
+	return
+}
+
+func CheckBLSSummary(summary []byte) bool {
+	guardianAddr, blsPubkey, blsPop, holderSig, ok := parseBLSSummary(summary)
+	if !ok {
+		return false
+	}
+
+	return checkBlsSummary(blsPubkey, blsPop, holderSig, guardianAddr)
+}
+
+func checkBlsSummary(blsPubkey *bls.PublicKey, blsPop *bls.Signature, holderSig *crypto.Signature, guardianAddr common.Address) bool {
+	if blsPubkey.IsEmpty() {
+		return false
+	}
+	if blsPop.IsEmpty() {
+		return false
+	}
+	if holderSig == nil || holderSig.IsEmpty() {
+		return false
+	}
+
+	if !holderSig.Verify(blsPop.ToBytes(), guardianAddr) {
+		return false
+	}
+
+	if !blsPop.PopVerify(blsPubkey) {
+		return false
+	}
+
+	return true
+}
+
+// StakeToGuardian stake Theta to given guardian node.
+func StakeToGuardian(db StateDB, sender common.Address, guardianSummary []byte, amount *big.Int) bool {
+	if amount.Cmp(core.MinGuardianStakeDeposit) < 0 {
+		return false
+	}
+	if db.GetThetaBalance(sender).Cmp(amount) < 0 {
+		return false
+	}
+
+	guardianAddr, blsPubkey, blsPop, holderSig, ok := parseBLSSummary(guardianSummary)
+	if !ok {
+		return false
+	}
+
+	view := db.(*state.StoreView)
+	gcp := view.GetGuardianCandidatePool()
+	if !gcp.Contains(guardianAddr) {
+		if !checkBlsSummary(blsPubkey, blsPop, holderSig, guardianAddr) {
+			return false
+		}
+	}
+
+	db.SubThetaBalance(sender, amount)
+
+	err := gcp.DepositStake(sender, guardianAddr, amount, blsPubkey, view.GetBlockHeight())
+	if err != nil {
+		return false
+	}
+
+	view.UpdateGuardianCandidatePool(gcp)
+
+	return true
+}
+
+// UnstakeFromGuardian unstake from Guardians.
+func UnstakeFromGuardian(db StateDB, addr common.Address, guardianAddr common.Address) bool {
+	view := db.(*state.StoreView)
+	gcp := view.GetGuardianCandidatePool()
+	currentHeight := view.GetBlockHeight()
+	err := gcp.WithdrawStake(addr, guardianAddr, currentHeight)
+	if err != nil {
+		return false
+	}
+
+	view.UpdateGuardianCandidatePool(gcp)
+	return true
+}
+
+// StakeToEEN stake to given EEN node.
+func StakeToEEN(db StateDB, sender common.Address, summary []byte, amount *big.Int) bool {
+	minEliteEdgeNodeStake := core.MinEliteEdgeNodeStakeDeposit
+	maxEliteEdgeNodeStake := core.MaxEliteEdgeNodeStakeDeposit
+
+	if amount.Cmp(minEliteEdgeNodeStake) < 0 {
+		return false
+	}
+
+	eenAddr, blsPubkey, blsPop, holderSig, ok := parseBLSSummary(summary)
+	if !ok {
+		return false
+	}
+
+	view := db.(*state.StoreView)
+
+	currentStake := big.NewInt(0)
+
+	eenp := state.NewEliteEdgeNodePool(view, false)
+	een := eenp.Get(eenAddr)
+	if een != nil {
+		currentStake = een.TotalStake()
+	}
+
+	expectedStake := big.NewInt(0).Add(currentStake, amount)
+	if expectedStake.Cmp(maxEliteEdgeNodeStake) > 0 {
+		return false
+	}
+
+	if db.GetBalance(sender).Cmp(amount) < 0 {
+		return false
+	}
+
+	if !checkBlsSummary(blsPubkey, blsPop, holderSig, eenAddr) {
```

Comment:
In tx_deposit_stake.go seems we only check the BLS summary when the current EENP doesn't contain the holder address. Here we want to be more conservative?

cf.: https://github.com/thetatoken/theta-protocol-ledger/blob/583dea58dffe61691a593238145038a63bdc738e/ledger/execution/tx_deposit_stake.go#L196

Reply:
updated